### DISCUSSION TITLE
React TypeError `setCell` is not a function when assigning function through `useStore`

### DISCUSSION THREAD

**Discussion by AdamGEmerson** [pmndrs/zustand#2215] [Q&A] [answered]
## Summary
I have an issue where functions in my store are not being properly imported into child components.  If anyone could take a look I would be greatly appreciative.  I can't figure out if this a bug, an issue in my setup, or a fundamental misunderstanding of how Zustand works...

## Versions
`node 21.2.0` 
`typescript 5.3.2`
`zustand 4.4.6`
`react 18.2`

## Expected Behavior
I am generating a simple crossword style grid and when I right click on a cell it should be converted to a black cell through an update to the grid state.

## Actual Behavior
When I right click a cell, I get `Uncaught TypeError: setCell is not a function`, where setCell is the function imported from the Zustand store.  Oddly, `setCell` is imported the same way in `App.tsx` and works, but I receive this error in `CrosswordCell.tsx`  

## Reproducible Example
[CodeSandbox](https://codesandbox.io/p/github/AdamGEmerson/crossword/main?layout=%257B%2522sidebarPanel%2522%253A%2522GIT%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpe99ksn00073b6f1ngzkt1n%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpe99ksn00033b6fdmmeie6a%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpe99ksn00053b6fa3y7lcb3%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpe99ksn00063b6f3yj8no5m%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B41.851918062396365%252C58.148081937603635%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpe99ksn00033b6fdmmeie6a%2522%253A%257B%2522id%2522%253A%2522clpe99ksn00033b6fdmmeie6a%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clpe99ksn00063b6f3yj8no5m%2522%253A%257B%2522id%2522%253A%2522clpe99ksn00063b6f3yj8no5m%2522%252C%2522activeTabId%2522%253A%2522clpe9czr000qq3b6f64dnkgg7%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522clpe9czr000qq3b6f64dnkgg7%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clpe99ksn00053b6fa3y7lcb3%2522%253A%257B%2522id%2522%253A%2522clpe99ksn00053b6fa3y7lcb3%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpe99ksn00043b6f650fjnig%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clpe99kwu000redhw2x0d81sk%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clpe99lwd007d3b6ftvgt0psd%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522CSB_RUN_OUTSIDE_CONTAINER%253D1%2520devcontainer%2520templates%2520apply%2520--template-id%2520%255C%2522ghcr.io%252Fdevcontainers%252Ftemplates%252Ftypescript-node%255C%2522%2520--template-args%2520%27%257B%257D%27%2520--features%2520%27%255B%255D%27%2522%252C%2522id%2522%253A%2522clpe9a4ca00ae3b6frjqkeiac%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpe9a4ca00ae3b6frjqkeiac%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

## Code

```ts
type PuzzleState = {
    grid: Array<Array<GridCell>>,
    setGrid: (grid: GridCell[][]) => void,
    setCell: (i: number, j: number, cell: GridCell) => void,
  }
```

```ts
// src/stores/puzzle.ts

import {create} from "zustand";

export const usePuzzleStore = create<PuzzleState>()(
  (set) => ({
    grid: [[]],
    setGrid: (grid: Array<Array<GridCell>>) => set(() => ({grid: grid})),
    setCell: (i: number, j: number, cell: GridCell) => set((state) => {
      const tempGrid = state.grid;
      tempGrid[i][j] = cell;
      return {grid: tempGrid};
    }, true),
  })
)
```

```ts
// App.tsx

import {usePuzzleStore} from "./stores/puzzle.ts";

// ...
function App() {
  const grid = usePuzzleStore(state => state.grid);
  const setGrid = usePuzzleStore(state => state.setGrid);
  const setCell = usePuzzleStore(state => state.setCell);

// ...

  useEffect(() => {
      console.log('grid changed');
      if(typeof grid[0][0] === 'undefined') return;
      if (grid[0][0].isBlack) return;
      setCell(0, 0, {...grid[0][0], isBlack: true}) // Works as expected
      setCell(0, 1, {...grid[0][1], isBlack: true})
      console.log("Set cell", grid);
    }, [grid]);

// ...

  return (
        <div className={'flex flex-col justify-around items-center w-screen'}>
          <h1 className={'m-4'}>
            Crossword App
          </h1>
          <div className="card grid grid-cols-10 rounded-lg overflow-hidden ring-4 ring-gray-900">
            {grid && grid.map((row, i) => (
              row.map((_cell, j) => {
                return <CrosswordCell key={`${i}-${j}`} i={i} j={j} />
              })
            ))}
          </div>
        </div>
    )
}
```

```ts
// CrosswordCell.tsx

import {usePuzzleStore} from "../stores/puzzle";

// ...
function CrosswordCell({i, j}: CrosswordCellProps) {
  const grid= usePuzzleStore((state) => state.grid);
  const setCell = usePuzzleStore((state) => state.setCell);

  // ...
  
  function handleRightClick(e: MouseEvent){
      e.preventDefault();
      console.log('right click');
      setCell(i, j, {...grid[i][j], isBlack: true}) // ERROR: Is not a function.
      console.log(grid)
    }

    return (
        <div className={"h-16 w-16"}
             onContextMenu={handleRightClick}
        >
          {!cell.isBlack ?
            <>
              <div className={'w-full z-10 absolute m-2'}>{cell.clue?.id}</div>
              <input
                maxLength={1}
                onContextMenu={handleRightClick}
                className="appearance-none z-0 text-center h-full w-full bg-blue-50 border border-slate-800 overflow-hidden text-3xl text-black shadow caret-blue-50"
                value={value}
                placeholder={placeholder ?? ''}
                onFocus={handleCellFocus}
                onBlur={handleCellBlur}
                onChange={handleCellChange}
              />
            </>
          : <div className="h-16 w-16 bg-slate-800 border border-slate-800 overflow-hidden text-3xl text-black shadow caret-blue-50"/>}
        </div>
    );
}
```

---

<img width="741" alt="Screenshot 2023-11-25 at 11 52 05 AM" src="https://github.com/pmndrs/zustand/assets/35545129/ba6c5c9d-7329-4685-b624-45ba566e80d1">

**Comment by dbritto-dev** [comment, marked as answer]
@AdamGEmerson you are replacing the whole state instead of shallow merging the new state so you ended up removing any field or function that does not exist on the new state. In a nutshell just remove the second parameter (true).

> Note: `set(...)` merge the new state with the previous one, `set(..., true)` replace the previous state with the new one so be careful when you use that. You shouldn't the need to replace the whole state except when you want to reset your state

`set(...)` -> state { a: 5, b: 9 } -> set({ b: 4 }) -> state { a:5, b: 4 }

`set(..., true)` -> state { a: 5, b: 9 } -> set({ b: 4 }, true) -> state { b: 4 }

**Reply by AdamGEmerson** [reply]
Thank you! I definitely misunderstood how the `replace` flag works.
